File: Ejemplos/38-DesayunoAsincronoJava/src/main/java/dev/joseluisgs/async/DesayunoAsync.java
package dev.joseluisgs.async;

import dev.joseluisgs.alimentos.*;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;


public class DesayunoAsync {
    public static void main(String[] args) throws ExecutionException, InterruptedException {

        var timeInit = System.currentTimeMillis();

        // Cafe taza = servirCafe();
        var taza = servirCafe();
        System.out.println("El café está listo");

        // CompletableFuture<Huevo> huevosTask = freirHuevosAsync(2);
        var huevosTask = freirHuevosAsync(2);
        var beiconTask = freirBeiconAsync(3);
        var tostadaTask = prepararTostadaConMantequillaYMeremeladaAsync(2);

        // Esperar a que todas las tareas se completen
        CompletableFuture<Void> todasLasTareas = CompletableFuture.allOf(huevosTask, beiconTask, tostadaTask);

        // Manejar las tareas completadas
        todasLasTareas.thenRun(() -> {
            try {
                if (huevosTask.isDone()) {
                    System.out.println("Los huevos están listos");
                }
                if (beiconTask.isDone()) {
                    System.out.println("El beicon está listo");
                }
                if (tostadaTask.isDone()) {
                    System.out.println("Las tostadas están listas");
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).join();

        Zumo zumoNaranja = servirZumoNaranja();
        System.out.println("El zumo de naranja está listo");
        System.out.println("¡El desayuno está listo!");

        var timeEnd = System.currentTimeMillis();
        System.out.println("Tiempo total: " + (timeEnd - timeInit) + " ms");
    }

    static CompletableFuture<Tostada> prepararTostadaConMantequillaYMeremeladaAsync(int rebanadas) {
        // Composición de acciones sincronas una vez completada la tarea
        return tostarPanAsync(rebanadas)
                .thenApply(tostada -> {
                    ponerMantequilla(tostada);
                    ponerMermelada(tostada);
                    return tostada;
                });
    }

    private static Zumo servirZumoNaranja() {
        System.out.println("Sirviendo zumo de naranja");
        return new Zumo();
    }

    private static void ponerMermelada(Tostada tostada) {
        System.out.println("Poniendo mermelada en la tostada");
    }

    private static void ponerMantequilla(Tostada tostada) {
        System.out.println("Poniendo mantequilla en la tostada");
    }

    private static CompletableFuture<Tostada> tostarPanAsync(int rebanadas) {
        return CompletableFuture.supplyAsync(() -> {
            for (int rebanada = 0; rebanada < rebanadas; rebanada++) {
                System.out.println("Poniendo una rebanada de pan en la tostadora");
            }
            System.out.println("Comenzando a tostar...");
            try {
                TimeUnit.SECONDS.sleep(3); // Simula la espera
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            System.out.println("Sacando la tostada de la tostadora");
            return new Tostada();
        });
    }

    private static CompletableFuture<Beicon> freirBeiconAsync(int lonchas) {
        return CompletableFuture.supplyAsync(() -> {
            System.out.println("Poniendo " + lonchas + " lonchas de beicon en la sartén");
            System.out.println("Cocinando la primera loncha del beicon...");
            try {
                TimeUnit.SECONDS.sleep(3); // Simula la espera
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            for (int rebanada = 0; rebanada < lonchas; rebanada++) {
                System.out.println("Dando la vuelta a una loncha de beicon");
            }
            System.out.println("Cocinando el segundo loncha del beicon...");
            try {
                TimeUnit.SECONDS.sleep(3); // Simula la espera
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            System.out.println("Poniendo el beicon en el plato");
            return new Beicon();
        });
    }

    private static CompletableFuture<Huevo> freirHuevosAsync(int cantidad) {
        return CompletableFuture.supplyAsync(() -> {
            System.out.println("Calentando la sartén para los huevos...");
            try {
                TimeUnit.SECONDS.sleep(3); // Simula la espera
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            System.out.println("Rompiendo " + cantidad + " huevos");
            System.out.println("Cocinando los huevos...");
            try {
                TimeUnit.SECONDS.sleep(3); // Simula la espera
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            System.out.println("Poniendo los huevos en el plato");
            return new Huevo();
        });
    }

    private static Cafe servirCafe() {
        System.out.println("Sirviendo café");
        return new Cafe();
    }
}

File: Ejemplos/37-DesayunoAsincronoKotlin/src/main/kotlin/async/DesayunoAsync.kt
package dev.joseluisgs.async

import dev.joseluisgs.alimentos.*
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking

fun main() = runBlocking {
    val timeInit = System.currentTimeMillis()

    val taza = servirCafe()
    println("El café está listo")

    // Lanzamos las corrutinas para las tareas asíncronas
    val huevosTask = async { freirHuevosAsync(2) }
    val beiconTask = async { freirBeiconAsync(3) }
    val tostadaTask = async { prepararTostadaConMantequillaYMermeladaAsync(2) }

    // Esperamos a que todas las tareas terminen
    val results = awaitAll(huevosTask, beiconTask, tostadaTask)

    results.forEach {
        when (it) {
            is Huevo -> println("Los huevos están listos")
            is Beicon -> println("El beicon está listo")
            is Tostada -> println("Las tostadas están listas")
        }
    }

    val zumoNaranja = servirZumoNaranja()
    println("El zumo de naranja está listo")
    println("¡El desayuno está listo!")

    val timeEnd = System.currentTimeMillis()
    println("Tiempo total: ${timeEnd - timeInit} ms")
}

private suspend fun prepararTostadaConMantequillaYMermeladaAsync(rebanadas: Int): Tostada {
    val tostada = tostarPanAsync(rebanadas) // no es necesario el await
    ponerMantequilla(tostada)
    ponerMermelada(tostada)
    return tostada
}

private fun servirZumoNaranja(): Zumo {
    println("Sirviendo zumo de naranja")
    return Zumo()
}

private fun ponerMermelada(tostada: Tostada) {
    println("Poniendo mermelada en la tostada")
}

private fun ponerMantequilla(tostada: Tostada) {
    println("Poniendo mantequilla en la tostada")
}

private suspend fun tostarPanAsync(rebanadas: Int): Tostada {
    for (rebanada in 0 until rebanadas) {
        println("Poniendo una rebanada de pan en la tostadora")
    }
    println("Comenzando a tostar...")
    delay(3000) // Simula la espera
    println("Sacando la tostada de la tostadora")
    return Tostada()
}

private suspend fun freirBeiconAsync(lonchas: Int): Beicon {
    println("Poniendo $lonchas lonchas de beicon en la sartén")
    println("Cocinando la primera loncha del beicon...")
    delay(3000) // Simula la espera
    for (rebanada in 0 until lonchas) {
        println("Dando la vuelta a una loncha de beicon")
    }
    println("Cocinando el segundo loncha del beicon...")
    delay(3000) // Simula la espera
    println("Poniendo el beicon en el plato")
    return Beicon()
}

private suspend fun freirHuevosAsync(cantidad: Int): Huevo {
    println("Calentando la sartén para los huevos...")
    delay(3000) // Simula la espera
    println("Rompiendo $cantidad huevos")
    println("Cocinando los huevos...")
    delay(3000) // Simula la espera
    println("Poniendo los huevos en el plato")
    return Huevo()
}

private fun servirCafe(): Cafe {
    println("Sirviendo café")
    return Cafe()
}